tests/model_fields/test_jsonfield.py

```diff
@@ -808,6 +808,46 @@ def test_deep_lookup_transform(self):
         )
         self.assertIs(NullableJSONModel.objects.filter(value__c__lt=5).exists(), False)
 
+    def test_lookups_special_chars(self):
+        value = {
+            "CONTROL": [],
+            "single'": [],
+            'double"': [
+                connection.features.supports_double_quotes_in_json_field_lookups,
+            ],
+            "dollar$": [],
+            "dot.dot": [],
+            "with space": [],
+            "back\\slash": [],
+            "question?mark": [],
+            "user@name": [],
+            "emo🤡'ji": [],
+            "com,ma": [],
+            "curly{{{brace}}}s": [],
+            "escape\uffff'seq'\uffffue\uffff'nce": [],
+            "m\\i@x. m🤡'a,t{{{ch}}}e?d$\"'es\uffff'ca\uffff'pe": [
+                connection.features.supports_double_quotes_in_json_field_lookups,
+            ],
+        }
+        obj = NullableJSONModel.objects.create(value=value)
+        obj.refresh_from_db()
+        self.assertEqual(obj.value, value)
+
+        for key, required_features in value.items():
+            lookups = {
+                "has_key": Q(value__has_key=key),
+                "has_keys": Q(value__has_keys=[key, "CONTROL"]),
+                "has_any_keys": Q(value__has_any_keys=[key, "does_not_exist"]),
+                "exact": Q(**{f"value__{key}": value[key]}),
+            }
+            for lookup, condition in lookups.items():
+                results = NullableJSONModel.objects.filter(condition)
+                with self.subTest(key=key, lookup=lookup):
+                    if all(required_features):
+                        self.assertSequenceEqual(results, [obj])
+                    else:
+                        self.assertSequenceEqual(results, [])
```

Comment:
This is quite complicated to understand, at least for me. I'd split them e.g.:
```python
    def test_lookups_special_chars(self):
        test_keys = [
            "CONTROL",
            "single'",
            "dollar$",
            "dot.dot",
            "with space",
            "back\\slash",
            "question?mark",
            "user@name",
            "emo🤡'ji",
            "com,ma",
            "curly{{{brace}}}s",
            "escape\uffff'seq'\uffffue\uffff'nce",
        ]
        json_value = {key: "some value" for key in test_keys}
        obj = NullableJSONModel.objects.create(value=json_value)
        obj.refresh_from_db()
        self.assertEqual(obj.value, json_value)     
        
        for key in test_keys:
            lookups = {              
                "has_key": Q(value__has_key=key),
                "has_keys": Q(value__has_keys=[key, "CONTROL"]),
                "has_any_keys": Q(value__has_any_keys=[key, "does_not_exist"]),
                "exact": Q(**{f"value__{key}": "some value"}),
            }                        
            for lookup, condition in lookups.items():
                results = NullableJSONModel.objects.filter(condition)
                with self.subTest(key=key, lookup=lookup):
                    self.assertSequenceEqual(results, [obj])
                        
    def test_lookups_special_chars_double_quotes(self):
        test_keys = [
            'double"',
            "m\\i@x. m🤡'a,t{{{ch}}}e?d$\"'es\uffff'ca\uffff'pe",
        ]
        json_value = {key: "some value" for key in test_keys}
        obj = NullableJSONModel.objects.create(value=json_value)
        obj.refresh_from_db()
        self.assertEqual(obj.value, json_value)
        self.assertSequenceEqual(NullableJSONModel.objects.filter(value__has_keys=test_keys), [obj])
        for key in test_keys:                                                    
            with self.subTest(key=key):
                results = NullableJSONModel.objects.filter(                      
                    Q(value__has_key=key),
                    Q(value__has_any_keys=[key, "does_not_exist"]),
                    Q(**{f"value__{key}": "some value"}),
                )
                self.assertSequenceEqual(results, [obj])

```

Reply:
Thank you, that looks great! Yeah I couldn't get the `PASSING` clause to work so I considered adding another `DatabaseFeature` flag for the single quotes specifically, thus structured the test in a way that would make it easy to test multiple flags – but that's not necessary anymore.

I've split the tests as per your suggestion 👍